index/segment/mem/results_iter.go

```diff
@@ -0,0 +1,144 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+)
+
+type resultsIter struct {
+	current document
+	next    document
+	hasNext bool
+	done    bool
+	err     error
+
+	predicateFn  matchPredicate
+	postingsList segment.ImmutablePostingsList
+	idsIter      segment.PostingsIter
+	queryable    queryable
+}
+
+func newResultsIter(
+	postingsList segment.ImmutablePostingsList,
+	predicateFn matchPredicate,
+	q queryable,
+) segment.ResultsIter {
+	// return empty iterator if we were given an empty postings list
+	if postingsList == nil {
+		return &resultsIter{
+			done: true,
+		}
+	}
+
+	it := &resultsIter{
+		predicateFn:  predicateFn,
+		postingsList: postingsList,
+		idsIter:      postingsList.Iter(),
+		queryable:    q,
+	}
+	it.setupNextIteration()
+	return it
+}
+
+func (r *resultsIter) setupNextIteration() {
+	d, ok, err := r.getNextDocument()
+	if err != nil {
+		r.release()
+		r.hasNext = false
+		r.err = err
+		r.done = true
+		return
+	}
+
+	if !ok {
+		r.release()
+		r.hasNext = false
+		r.done = true
+		return
+	}
+
+	// setup values so that next time Next() is called, we can rotate
+	// next -> current
+	r.next = d
+	r.hasNext = true
+}
+
+func (r *resultsIter) release() {
+	// need this cast because we're provided an immutable postings lists
+	// and we pool mutable postings lists
+	pl, ok := r.postingsList.(segment.PostingsList)
+	if ok {
+		r.queryable.Options().PostingsListPool().Put(pl)
+		// unset internal references
+		r.postingsList = nil
+		r.idsIter = nil
+	}
+}
+
+func (r *resultsIter) getNextDocument() (document, bool, error) {
+	// need to loop as we may filter out values due to the predicate function
+	for {
+		hasNext := r.idsIter.Next()
+		if !hasNext {
+			return document{}, false, nil
+		}
+
+		id := r.idsIter.Current()
+		doc, ok := r.queryable.FetchDocument(id)
```

Comment:
Feels weird that we actually iterate but then do a hash lookup per each document we iterate over? (when using a simple segment)

Reply:
I have a TODO in there to see impact of making `FetchDocument` slice backed instead of hash backed. Will that might help address the concern?

The main reason for the is it lets us do the intersection cheaply across filters by using a 
postings list of the ints. Probably will retain a `filter -> postingsList` cache in the near term that'll need this too.